File: source/python/1129. Shortest Path with Alternating Colors.py
from queue import Queue


class Solution:
    def shortestAlternatingPaths(
        self, n: int, redEdges: list[list[int]], blueEdges: list[list[int]]
    ) -> list[int]:
        if n == 1:
            return [0]

        # Create adjacency lists for red and blue edges
        red: list[list[int]] = [[] for _ in range(n)]
        blue: list[list[int]] = [[] for _ in range(n)]

        # Initialize the adjacency lists
        for u, v in redEdges:
            red[u].append(v)

        for u, v in blueEdges:
            blue[u].append(v)

        # Initialize the queue and put the starting node in it
        q: Queue[tuple[int, str, int]] = Queue()
        q.put((0, 'r', 0))
        q.put((0, 'b', 0))

        # Initialize the visited set and put the starting node in it
        visited: set[tuple[int, str]] = set()
        visited.add((0, 'r'))
        visited.add((0, 'b'))

        # Create the answer array and initialize the first element
        ans = [-1] * n
        ans[0] = 0

        # Traverse the graph
        while not q.empty():
            u, color, d = q.get()

            # If the current color is red, traverse the blue edges, and vice versa
            if color == 'r':
                for v in red[u]:
                    # If the node has already been visited with the opposite color, continue
                    if (v, 'b') in visited:
                        continue

                    # If the node has not been visited, add it to the visited set and the queue
                    visited.add((v, 'b'))
                    q.put((v, 'b', d + 1))

                    # If the node has not been visited before, update the answer array
                    if ans[v] == -1:
                        ans[v] = d + 1
            else:
                for v in blue[u]:
                    if (v, 'r') in visited:
                        continue

                    visited.add((v, 'r'))
                    q.put((v, 'r', d + 1))

                    if ans[v] == -1:
                        ans[v] = d + 1

        return ans


def test():
    s = Solution()

    assert s.shortestAlternatingPaths(
        3, [[0, 1], [1, 2]], [[1, 0]]
    ) == [0, 1, -1]
    assert s.shortestAlternatingPaths(3, [[0, 1]], [[2, 1]]) == [0, 1, -1]

    print("All tests passed!")


if __name__ == "__main__":
    test()

File: source/cpp/1129. Shortest Path with Alternating Colors.cpp
#include <queue>
#include <vector>


class Solution {
public:
    std::vector<int> shortestAlternatingPaths(
        const int& n, std::vector<std::vector<int>>& redEdges, std::vector<std::vector<int>>& blueEdges) {
        std::vector<std::vector<std::pair<int, int>>> adj(n);
        for (auto& redEdge : redEdges) adj[redEdge[0]].push_back({redEdge[1], 0});
        for (auto& blueEdge : blueEdges) adj[blueEdge[0]].push_back(std::make_pair(blueEdge[1], 1));

        std::vector<int> answer(n, -1);
        std::vector<std::vector<bool>> visit(n, std::vector<bool>(2));
        std::queue<std::tuple<int, int, int>> q;

        // Start with node 0, with number of steps as 0 and undefined color -1.
        q.push({0, 0, -1});
        visit[0][1] = visit[0][0] = true;
        answer[0] = 0;

        while (!q.empty()) {
            const auto [node, steps, prevColor] = q.front();
            q.pop();

            for (const auto& [neighbor, color] : adj[node]) {
                if (color == prevColor || visit[neighbor][color] == true) continue;
                
                visit[neighbor][color] = true;
                q.push({neighbor, 1 + steps, color});
                if (answer[neighbor] == -1) answer[neighbor] = 1 + steps;
            }
        }

        return answer;
    }
};
